Min # Vending step5

undervineg opened the pull request.

undervine committed:
Modify contents

undervine committed:
Replace enum Notifcations to extension of Notification.Name

undervine committed:
Add contents about observer pattern and loosed coupling

undervine committed:
Add links

godrm commented:
이미 노티피케이션센터를 적용한 상태가 수정사항이 많지 않네요.
그러면 지난 단계에서 나왔던 싱글톤 객체 인스턴스에 의존하지 않도록 개선해봅시다.

undervineg commented:
sharedInstance에 접근할 때 프로토콜로 접근하면 될까요?

```
protocol Singleton {
    static func shared() -> Self
    static func restoreStates(_ machine: Self)
}

```
```
final class VendingMachine: Sequence, Machine, Singleton {
    static func shared() -> VendingMachine {
        return sharedInstance
    }

    static var sharedInstance: VendingMachine {
        return VendingMachine()
    }
}
```

godrm commented:
싱글톤 객체를 위해 프로토콜을 바꾸는게 아니라, 
ViewController가 sharedInstance()로 직접 접근하지 않도록 의존성을 외부에서 주입하는 방식을 고민해보세요.

undervine committed:
Remove protocols - Machine, Product, and Purchasable

Remove some generic protocols because it's not possible to load storyboard from ViewControllers using generics.

undervine committed:
Make ViewController to get VendingMachine object, instead of accessing sharedInstance of VendingMachine

Make ViewController independent of Model-VendingMachine.

undervineg commented:
- 제네릭 프로토콜 모두 제거
- ViewController의 모델 객체(machine)를 프로토콜 타입으로 선언
- AppDelegated에서 ViewController에 VendingMachine 주입
- MachineStore를 VendingMachine 파일로 옮김 -> VendingMachine의 shared()를 fileprivate으로 지정하여 VendingMachine 객체는 MachineStore에서만 접근 가능하게 함

godrm commented:
우어 엄청난 대공사였네요. 제네릭 프로토콜 타입을 사용한게 UIViewController 에서 활용할 때 문제가 되니까 주의할 필요가 있습니다.
수고하셨습니다. 의존성이 이렇게 무섭습니다 😱
다음 단계로 갑시다.

undervineg commented:
ㅎㅎㅎ 네~!

undervineg commented:
JK 머지 해도 되나요? 아직 머징됐다는 표시가 안 뜨는데..

undervine committed:
Add contents about dependency injection

godrm commented:
앗.. 머지를 안 눌렀네요 ㅜㅜ 네 갑시다.
